In [1]:
# Importing the neccessary libraries
import pandas as pd
import numpy as np
import statistics as stats
import os
import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from datetime import datetime

# Forcing keras to use CPU.
os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

In [2]:
# Reading the Data and storing it in a dataframe

df = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
df.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

In [3]:
print('\nShape of dataframe : ',df.shape)


Shape of dataframe :  (1030, 9)


In [4]:
# Summary of the dataset
df.describe()

Cement  Blast Furnace Slag  ...          Age     Strength
count  1030.000000         1030.000000  ...  1030.000000  1030.000000
mean    281.167864           73.895825  ...    45.662136    35.817961
std     104.506364           86.279342  ...    63.169912    16.705742
min     102.000000            0.000000  ...     1.000000     2.330000
25%     192.375000            0.000000  ...     7.000000    23.710000
50%     272.900000           22.000000  ...    28.000000    34.445000
75%     350.000000          142.950000  ...    56.000000    46.135000
max     540.000000          359.400000  ...   365.000000    82.600000

[8 rows x 9 columns]

In [5]:
# Creating dataframes of features (X) and target (Y)
X = df.iloc[:, 0:8]
Y = df.iloc[:,8]

# Printing the dataframes X and Y to ensure we have created the dataframes with the correct columns
print('The features or the predictors (X) are : ', X, '\n\n') 
print('The target (Y) is : ', Y, '\n\n')

The features or the predictors (X) are :        Cement  Blast Furnace Slag  ...  Fine Aggregate  Age
0      540.0                 0.0  ...           676.0   28
1      540.0                 0.0  ...           676.0   28
2      332.5               142.5  ...           594.0  270
3      332.5               142.5  ...           594.0  365
4      198.6               132.4  ...           825.5  360
...      ...                 ...  ...             ...  ...
1025   276.4               116.0  ...           768.3   28
1026   322.2                 0.0  ...           813.4   28
1027   148.5               139.4  ...           780.0   28
1028   159.1               186.7  ...           788.9   28
1029   260.9               100.5  ...           761.5   28

[1030 rows x 8 columns] 


The target (Y) is :  0       79.99
1       61.89
2       40.27
3       41.05
4       44.30
        ...  
1025    44.28
1026    31.18
1027    23.70
1028    32.77
1029    32.40
Name: Strength, Length: 1030, dtype: float64 



<b>Note 1</b> : Unlike the method in this course, the splitting is done using indexing instead of using the names of the columns. Additionally, a different notation is used. The word <i>features</i> is used instead of <i>predictors</i>.


<b>Note 2</b> : Pandas indexes columns starting from 0. Note in the code below for the features (X) indexing is used as `[:, 0:8]`. The first part preceding the coma `(:)` tells pandas to include ALL rows of the original dataframe (df) in the new dataframe called X while the part succedding the comma `(0:8)` tells pandas to include all columns of the original dataframe (df) starting from column with index = 0 and ending with column with index = 7, <b> but not to include the column with index = 8 </b>  

<b>Note 3</b> : In order to split the data into train and test sets, the train_test_split function of the sklearn library is used. `The random_state` is used to ensure that the train and test split is the same each time, i.e. the train set and the test set have the same samples each time the code is run which is good for reproducing the results. If left empty, the random state is used by `np.random`. Since the Project requires splitting data the into <b>random</b> sets, hence `random_state` is not used, i.e. no value is set for random state. As the data hase to be split randomly  into train and test sets <b>50</b> times, a for loop will be used to to split the data in train test sets for <b>each model</b> 

In [6]:
def regression_model() :
    
    # Create the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(X.shape[1],)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))

    # Compile the model
    model.compile(optimizer='adam', loss='mean_squared_error')
    
    return model

In [7]:
def data_split() :
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3)
    
    # Create a list containing X_train, X_test, Y_train, Y_test and return the list
    splits = [X_train, X_test, Y_train, Y_test] 
    return splits

<b>Note </b> : In the above function `data_split()` the <i>X_train, X_test, Y_train, Y_test</i> sets are stored in a list and list is returned. This is to ensure that the <i>X_train, X_test, Y_train, Y_test</i> sets are not printed when the function is called

In [8]:
def predict() :
    return model.predict(X_test)

def calculate_mse() :
    return mean_squared_error(Y_test,Y_predicted)

<b>Note 1</b> : The function `regression_model` as defined above only **compiles** the model and doesn't fit the model to training set. This is because in PART C, the number of epochs are changed and this would allow to fit the model with new number of epochs. This function will be used for **PART A**, **PART B** and **PART C**.

<b>Note 2</b> : Since the splitting data, prediction and mean squared error calculations are all being used repeatly as well, hence there are separate functions created for them to remove redundancy of typing the same line of code while also making the code more neat. However,  as we the features (X) are to be normalized only **once**, hence there is no need to create a function for it

<b>Note 3</b> : As the split function is executed first, the training and test sets are obtained and hence there is no need to explicitly pass any arguments to `predict()` and `calculate_mse()` functions. 

# <font color = blue> PART B : BASELINE MODEL WITH NORMALIZED FEATURES </font>


In this part, all the tasks from <b>PART A</b> are performed, but this time the values for the features (X) will be normalized using the formula `X - µ / σ`



## <font color = darkorange>Task 1 : Train and Test the Baseline Model with Normalized Features</font>

In order to train and test the the baseline model with normalized features, the following steps are performed :
<ol>
    <li>Normalize the features (X)</li>
    <li>Randomly split the data into <i>X_train, X_test, Y_train, Y_test</i> sets</li>
    <li>Train the model using <i>X_train, Y_train</i> using <b>50</b> epochs</li>
    <li>Get the predictions on the <i>X_test</i> set </li>
    <li>Compute the <b>mean squared error</b> on the test set using sklearn</li>
</ol>

### <font color = #2980B9> Step 1 : Normalize the features (X) </font>

In [9]:
X = (X - X.mean()) / X.std()

### <font color = #2980B9> Step 2 : Randomly split the data into <i>X_train, X_test, Y_train, Y_test </font>

In [10]:
# Creating X_train, X_test, Y_train and Y_test sets
X_train, X_test, Y_train, Y_test = data_split()

### <font color = #2980B9> Step 3 : Train the model using <i>X_train, Y_train</i> using 50 epochs </font>

<b>Note </b> : Since the regression model has already been created using `def regression_model` in <b>TASK 1</b> and there are no changes being made to it, hence there is no need to write the code for the model again. 

In [11]:
model = regression_model()

# Fit the model on the train set
model.fit(X_train, Y_train, validation_split=0.3, epochs=50)

Epoch 1/50
16/16 [==============================] - 1s 36ms/step - loss: 1617.6201 - val_loss: 1492.4595
Epoch 2/50
16/16 [==============================] - 0s 4ms/step - loss: 1558.0160 - val_loss: 1478.9688
Epoch 3/50
16/16 [==============================] - 0s 4ms/step - loss: 1476.0178 - val_loss: 1464.9487
Epoch 4/50
16/16 [==============================] - 0s 4ms/step - loss: 1524.2386 - val_loss: 1449.4432
Epoch 5/50
16/16 [==============================] - 0s 4ms/step - loss: 1417.0921 - val_loss: 1432.4918
Epoch 6/50
16/16 [==============================] - 0s 4ms/step - loss: 1489.7961 - val_loss: 1412.7939
Epoch 7/50
16/16 [==============================] - 0s 4ms/step - loss: 1442.0594 - val_loss: 1390.7305
Epoch 8/50
16/16 [==============================] - 0s 4ms/step - loss: 1407.4251 - val_loss: 1364.8779
Epoch 9/50
16/16 [==============================] - 0s 4ms/step - loss: 1407.9444 - val_loss: 1334.6074
Epoch 10/50
16/16 [==============================] - 0s 4ms/ste

### <font color = #2980B9> Step 4 : Get the predictions on the X_test </font>

In [12]:
# Store the predictions in a variable Y_Predicted
Y_predicted = predict()

### <font color = #2980B9> Step 5 : Compute the <i>mean squared error</i> on the test set using sklearn </font>

In [13]:
# Calculate the mean square error

mse = calculate_mse()
print('Mean Square Error (MSE) of the Baseline Model with Normalized Features is : ' , str(mse))

Mean Square Error (MSE) of the Baseline Model with Normalized Features is :  178.43257144321913


## <font color = darkorange>Task 2 : Create 50 Models and Calculate the <i>µ</i> & <i>σ</i> of their MSE Errors with new Features</font>

In order to train 50 models with new features (normalized features) and calulate the mean (µ) and standard deviation (σ) of their mean square errors (MSE) the following steps are performed :
<ol>
    <li>Create an empty list <code>list_of_mse</code> to store the mean square error of each of the models</li>
    <li>Define a for loop and perform each of the following steps in the loop</li>
        <ol>
        <li>Randomly split the data into <i>X_train, X_test, Y_train, Y_test</i> sets</li>
        <li>Train the model using <i>X_train, Y_train</i> using <b>50</b> epochs</li>
        <li>Evaluate the model using <i>X_test, Y_test</i></li>
        <li>Get the predictions on the <i>X_test</i> set </li>
        <li>Compute the <b>mean squared error</b> on the test set using sklearn</li>
    </ol>
</ol>

### <font color = #2980B9> Step 1 : Creating the <i>list_of_mse</i> list</font>

In [14]:
# Create the empty lists
list_of_mse = []

### <font color = #2980B9> Step 2 : Creating 50 Models, Calculating The MSE for Each and <i>µ</i> & <i>σ</i> of the 50 MSE Values in <i>list_of_mse</i> </font>

In [15]:
# Create the for loop to split the data, create, compile & fit model, evaluate & nake predictions, caluclate mse and store
# in list_of_mse

start_time = datetime.now() # Starting time of the for loop execution

for i in range(50) :
    # Split the data into train and test set
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3)
    
    # Create and compile the regression model using the function regression_model as defined in TASK 1
    model = regression_model()

    # Fit the model on the train set
    print('\n\n\nTraining Model # ' , i+1 , '\n\n') # Print the Model Number that is being trained
    model.fit(X_train, Y_train, validation_split=0.3, epochs=50)
    print('\n')
    
    # Make prediction on the test set
    Y_predicted = model.predict(X_test)
    
    # Calculate the mean square error
    mse = mean_squared_error(Y_test, Y_predicted)
    
    # Add the mse to the list_of_mse list
    list_of_mse.append(mse)

end_time = datetime.now() # Ending time of the for loop execution

# Print time taken for fitting 50 models and calucating the Mean and Standard Deviation of MSE of 50 models
print('\n\nTotal Execution Time : ' , format(end_time - start_time))
    

Streaming output truncated to the last 5000 lines.
16/16 [==============================] - 0s 4ms/step - loss: 206.5160 - val_loss: 192.4452
Epoch 38/50
16/16 [==============================] - 0s 4ms/step - loss: 197.8638 - val_loss: 190.0580
Epoch 39/50
16/16 [==============================] - 0s 5ms/step - loss: 192.0201 - val_loss: 188.1503
Epoch 40/50
16/16 [==============================] - 0s 4ms/step - loss: 201.8564 - val_loss: 186.4283
Epoch 41/50
16/16 [==============================] - 0s 5ms/step - loss: 184.7096 - val_loss: 184.3759
Epoch 42/50
16/16 [==============================] - 0s 5ms/step - loss: 170.8807 - val_loss: 182.6363
Epoch 43/50
16/16 [==============================] - 0s 5ms/step - loss: 185.3382 - val_loss: 181.2137
Epoch 44/50
16/16 [==============================] - 0s 4ms/step - loss: 181.3650 - val_loss: 179.4454
Epoch 45/50
16/16 [==============================] - 0s 5ms/step - loss: 172.6410 - val_loss: 177.7108
Epoch 46/50
16/16 [===============

In [16]:
# Calculate the Mean of the MSE of 50 models
mean_of_mse = stats.mean(list_of_mse)

# Calculate the Standard Deviation of the MSE of 50 models
std_of_mse = stats.stdev(list_of_mse)

# Print the Mean and Standard Deviation of MSE of 50 models
print('\n\nMean of the MSE of 50 Models : ' , str(mean_of_mse))
print('Standard Deviation of MSE of 50 Models : ' + str(std_of_mse))



Mean of the MSE of 50 Models :  176.2670903912207
Standard Deviation of MSE of 50 Models : 15.40801984587299


#### <font color = green> Comparision of Mean of MSE with Mean of MSE from PART A </font>
<p/>

<table style="width:20%">
  <tr>
    <th>Mean of MSE of PART A</th>
    <th>Mean of MSE of PART B</th>
  </tr>
  <tr>
    <td>177.27</td>
    <td>176.27</td>
  </tr>
</table>

The table above compares the Mean of **MSE for PART A** and **Mean of MSE for PART B**. As can be seen, the value of Mean of MSE of PART B is significantly smaller than that of PART A. The mean squared error tells how close a regression line is to a set of points. The smaller the value, the closer the model is to finding the line of best fit. So the smaller value for PART B shows that the model is slightly closer to finding a line of best fit.

To summarize, normalizing the features had a significant effect in reducing the MSE and finding the line of best fit

<b>Note</b> : Depending on the data, it may be impossible to get a very small value for MSE.